### PR TITLE
Add SIS Estimate Functionality

### LINKED ISSUE #25: Estimate SIS and variants
- lattice attack (q-ary lattices)
- BKW style (optional)
- Arora-GB style (optional)

### PR DESCRIPTION
**Pull request by hkippen-SBAQ**
Resolves #25 

Adds a new, separate API for SIS estimation specifically, alongside a new SISParameters data class. Initial support is only for lattice attacks on SIS, so only an sis_lattice attack file is included. Outputs for SIS with an Infinity norm length bound are tuned to match the Dilithium NIST Round 3 specification https://pq-crystals.org/dilithium/data/dilithium-specification-round3-20210208.pdf when using the new basis shape estimator. Schemes.py has been updated to include unforgeability parameters for Dilithium and secret key recovery plus unforgeability parameters for Falcon.

### CODE REVIEW

File: docs/algorithms/sis-lattice.rst [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+SIS Lattice Attacks
+=====================
+
+We construct an (easy) example SIS instance::
+
+    from estimator import *
+    params = SIS.Parameters(n=113, q=2048, length_bound=512, norm=2) 
+    params
+
+The simplest (and quickest to estimate) model is solving for the SIS instance with a euclidian norm length bound and assuming the Gaussian heuristic [CheNgu12]_.Then, we can solve for the required root hermite factor [EC:GamNgu08]_ that will guarantee BKZ outputs a short enough vector::
```
Comment (malb) [inline docs/algorithms/sis-lattice.rst:12]: Why are we referencing CheNgu12 here?

File: docs/algorithms/sis-lattice.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    SIS.lattice(params)
+
+The exact reduction shape model doesn't matter when using euclidian norm bounds, as the required block size is calculated directly from the length bound.
```
Comment (malb) [inline docs/algorithms/sis-lattice.rst:16]: -> "does not"

File: docs/algorithms/sis-lattice.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    SIS.lattice(params.updated(length_bound=70), red_shape_model=Simulator.ZGSA)
+
+To get a more precise answer we may use the CN11 simulator by Chen and Nguyen [AC:CheNgu11]_ (as `implemented in FPyLLL <https://github.com/fplll/fpylll/blob/master/src/fpylll/tools/bkz_simulator.py>`__)::
+
+    SIS.lattice(params.updated(length_bound=70), red_shape_model=Simulator.CN11)
+
+Another option is to simulate a rerandomization of the basis, such that the q-vectors are *forgotten*. This results in the ``LGSA`` simulator, where the short, unit vectors are still present in the basis. See Figure 12 in the dilithium submission for an example.We can then improve on this result by first preprocessing the basis with block size β followed by a single SVP call in dimension η [RSA:LiuNgu13]_. We call this the BDD approach since this is essentially the same strategy as preprocessing a basis and then running a CVP solver::
```
Comment (malb) [inline docs/algorithms/sis-lattice.rst:31]: Capitalise "Dilithium"

Comment (malb) [reply docs/algorithms/sis-lattice.rst:31]: Space after Dilithium

Comment (malb) [reply docs/algorithms/sis-lattice.rst:31]: This BDD stuff seems a leftover from somewhere else?

File: estimator/sis_lattice.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        params: SISParameters,
+        ignore_qary: bool = False,
+        red_shape_model=red_simulator_default,
+        red_cost_model=red_cost_model_default,
+        d=None,
+        log_level=5,
+        **kwds,
+    ):
+        """
+        This function optimizes costs for a fixed guessing dimension ζ.
```
Comment (malb) [inline estimator/sis_lattice.py:180]: Guessing dimension might be a bit misleading?

**Review by malb — COMMENTED** [review]
Can we also add a doctest to README.rst to showcase the new capabilities?

File: docs/algorithms/sis-lattice.rst (hunk 1 above)
Comment (hunterkipt) [reply docs/algorithms/sis-lattice.rst]: Fair enough. Couldn't exactly find a better reference for the Gaussian heuristic. Removed.

File: docs/algorithms/sis-lattice.rst (hunk 2 above)
Comment (hunterkipt) [reply docs/algorithms/sis-lattice.rst]: Done

File: docs/algorithms/sis-lattice.rst (hunk 3 above)
Comment (hunterkipt) [reply docs/algorithms/sis-lattice.rst]: It was. Forgot to delete. Resolved.

File: estimator/sis_lattice.py (hunk 4 above)
Comment (hunterkipt) [reply estimator/sis_lattice.py]: Changed to make a bit more descriptive.

File: docs/algorithms/sis-lattice.rst (hunk 1 above)
Comment (malb) [reply docs/algorithms/sis-lattice.rst]: Ah, sorry, I misread, yes, this is fine. Sorry for the noise

**Comment by malb** [comment]
Wooohoo!

**Pull request merged**